services/mqtt/service.go

```diff
@@ -0,0 +1,186 @@
+	AtLeastOnce
+	// guarantees delivery only once. Safest and slowest.
+	ExactlyOnce
+)
+
+// DefaultQuiesceTimeout is the duration the client will wait for outstanding
+// messages to be published before forcing a disconnection
+const DefaultQuiesceTimeout time.Duration = 250 * time.Millisecond
+
+type Service struct {
+	logger *log.Logger
+
+	configValue atomic.Value
+	client      pahomqtt.Client
+	token       pahomqtt.Token
+}
+
+func NewService(c Config, l *log.Logger) *Service {
+	s := &Service{
+		logger: l,
+	}
+	s.configValue.Store(c)
+	return s
+}
+
+func (s *Service) config() Config {
+	return s.configValue.Load().(Config)
+}
+
+func (s *Service) Open() error {
+	s.logger.Println("I! Starting MQTT service")
+	s.runClient()
+	return nil
+
+}
+
+func (s *Service) Close() error {
+	s.logger.Println("I! Stopping MQTT service")
+	s.stopClient()
+	return nil
+}
+
+func (s *Service) Alert(qos QoSLevel, retained bool, topic, message string) error {
+	s.client.Publish(topic, byte(qos), retained, message)
+	return nil
+}
+
+func (s *Service) Update(newConfig []interface{}) error {
+	if l := len(newConfig); l != 1 {
+		return fmt.Errorf("expected only one new config object, got %d", l)
+	}
+	if c, ok := newConfig[0].(Config); !ok {
+		return fmt.Errorf("expected config object to be of type %T, got %T", c, newConfig[0])
+	} else {
+		s.configValue.Store(c)
+		if err := s.reloadClient(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Service) Handler(c HandlerConfig, l *log.Logger) alert.Handler {
+	return &handler{
+		s:      s,
+		c:      c,
+		logger: l,
+	}
+}
+
+type HandlerConfig struct {
+	Topic    string
+	QoS      QoSLevel
+	Retained bool
+}
+
+type handler struct {
+	s      *Service
+	c      HandlerConfig
+	logger *log.Logger
+}
+
+func (h *handler) Handle(event alert.Event) {
+	if err := h.s.Alert(h.c.QoS, h.c.Retained, h.c.Topic, event.State.Message); err != nil {
+		h.logger.Println("E! failed to post message to MQTT broker", err)
+	}
+}
+
+func (s *Service) DefaultHandlerConfig() HandlerConfig {
+	c := s.config()
+	return HandlerConfig{
+		Topic:    c.DefaultTopic,
+		QoS:      c.DefaultQoS,
+		Retained: c.DefaultRetained,
+	}
+}
+
+func (s *Service) reloadClient() error {
+	if err := s.stopClient(); err != nil {
+		return err
+	}
+	if err := s.runClient(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Service) runClient() error {
+	c := s.config()
+	opts := pahomqtt.NewClientOptions()
+	opts.AddBroker(c.Broker())
+	opts.SetClientID(c.ClientID)
+	opts.SetUsername(c.Username)
+	opts.SetPassword(c.Password)
+
+	// Using a clean session forces the broker to dispose of client session
+	// information after disconnecting. Retention of this is useful for
+	// constrained clients.  Since Kapacitor is only publishing, it has no
+	// storage requirements and can reduce load on the broker by using a clean
+	// session.
+	opts.SetCleanSession(true)
+
+	s.client = pahomqtt.NewClient(opts)
```

Comment:
Since `s.client` is being modified here you will need to lock access to it or use the atomic.Value type like you do for the config struct.